ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/ForkIdManager.java

```diff
@@ -39,69 +36,58 @@
     checkNotNull(blockchain);
-    checkNotNull(forks);
+    checkNotNull(nonFilteredForks);
+    this.chainHeadSupplier = blockchain::getChainHeadBlockNumber;
     this.genesisHash = blockchain.getGenesisBlock().getHash();
     this.forkAndHashList = new ArrayList<>();
-    final Predicate<ForkId> legacyForkIdChecker =
-        createForkIDChecker(
-            blockchain,
-            genesisHash,
-            forks,
-            fs ->
-                fs.stream()
-                    .filter(fork -> fork > 0)
-                    .distinct()
-                    .collect(Collectors.toUnmodifiableList()),
-            forkAndHashList);
-    // if the fork list contains only zeros then we may be in a consortium/dev network
-    if (onlyZerosForkBlocks(forks)) {
-      this.forkIDCheckers = singletonList(forkId -> true);
-    } else {
-      final Predicate<ForkId> newForkIdChecker =
-          createForkIDChecker(
-              blockchain,
-              genesisHash,
-              forks,
-              fs -> fs.stream().distinct().collect(Collectors.toUnmodifiableList()),
-              new ArrayList<>());
-      this.forkIDCheckers = Arrays.asList(newForkIdChecker, legacyForkIdChecker);
-    }
-  }
-
-  private static Predicate<ForkId> createForkIDChecker(
-      final Blockchain blockchain,
-      final Hash genesisHash,
-      final List<Long> forks,
-      final Function<List<Long>, List<Long>> sanitizer,
-      final List<ForkId> forkIds) {
-    final List<Long> sanitizedForks = sanitizer.apply(forks);
-    final long forkNext = createForkIds(genesisHash, sanitizedForks, forkIds);
-    return eip2124(blockchain, forkNext, forkIds, highestKnownFork(sanitizedForks));
+    this.forks =
+        nonFilteredForks.stream()
+            .filter(fork -> fork > 0)
+            .distinct()
+            .sorted()
+            .collect(Collectors.toUnmodifiableList());
+    this.onlyZerosForkBlocks = nonFilteredForks.stream().allMatch(value -> 0L == value);
+    this.forkNext = createForkIds();
+    this.highestKnownFork = !forks.isEmpty() ? forks.get(forks.size() - 1) : 0L;
   }
 
-  private static boolean onlyZerosForkBlocks(final List<Long> forks) {
-    return forks.stream().allMatch(value -> 0L == value);
-  }
+  public ForkId computeForkId() {
+    final long head = chainHeadSupplier.getAsLong();
+    if (lastHead != 0 && head == lastHead && lastComputedForkId != null) {
+      return lastComputedForkId;
+    }
+    lastHead = head;
 
-  private static long highestKnownFork(final List<Long> forks) {
-    return !forks.isEmpty() ? forks.get(forks.size() - 1) : 0L;
+    final CRC32 crc = new CRC32();
+    long next = 0;
+    crc.update(genesisHash.toArray());
+    for (final Long fork : forks) {
+      if (fork <= head) {
+        updateCrc(crc, fork);
+        continue;
+      }
+      next = fork;
+      break;
+    }
+    lastComputedForkId = new ForkId(getCurrentCrcHash(crc), next);
+    return lastComputedForkId;
```

Comment:
Is there a way to get this from the `forkAndHashList`? If I understand correctly we just need to filter out the hashes that are for forks past the chain head, right?

Reply:
Maybe we can. Let me check.

Reply:
Well. We can

Reply:
Fixed here https://github.com/hyperledger/besu/pull/1050/commits/db09f7677652d3ad561b2f78746b98e796f277ad